bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/mirroring/TestMirroringTable.java

```diff
@@ -0,0 +1,1281 @@
+        }
+      }
+    }
+    databaseHelpers.verifyTableConsistency(tableName1, failPredicate);
+
+    reportedErrorsContext1.assertNewErrorsReported(databaseEntriesCount / 2);
+
+    ReportedErrorsContext reportedErrorsContext2 = new ReportedErrorsContext();
+    final TableName tableName2 = connectionRule.createTable(columnFamily1);
+    try (MirroringConnection connection = databaseHelpers.createConnection()) {
+      try (Table t1 = connection.getTable(tableName2)) {
+        int id = 0;
+        for (int i = 0; i < databaseEntriesCount / 100; i++) {
+          List<Put> puts = new ArrayList<>();
+          for (int j = 0; j < 100; j++) {
+            puts.add(Helpers.createPut(id, columnFamily1, qualifier1));
+            id++;
+          }
+          t1.put(puts);
+        }
+      }
+    }
+    databaseHelpers.verifyTableConsistency(tableName2, failPredicate);
+    reportedErrorsContext2.assertNewErrorsReported(databaseEntriesCount / 2);
+  }
+
+  @Test
+  public void testDelete() throws IOException {
+    int databaseEntriesCount = 1000;
+    final TableName tableName1 = connectionRule.createTable(columnFamily1);
+    databaseHelpers.fillTable(tableName1, databaseEntriesCount, columnFamily1, qualifier1);
+    try (MirroringConnection connection = databaseHelpers.createConnection()) {
+      try (Table t1 = connection.getTable(tableName1)) {
+        for (int i = 0; i < databaseEntriesCount; i++) {
+          byte[] rowKey = rowKeyFromId(i);
+          t1.delete(Helpers.createDelete(rowKey, columnFamily1, qualifier1));
+        }
+      }
+    }
+    databaseHelpers.verifyTableConsistency(tableName1);
+
+    final TableName tableName2 = connectionRule.createTable(columnFamily1);
+    databaseHelpers.fillTable(tableName2, databaseEntriesCount, columnFamily1, qualifier1);
+    try (MirroringConnection connection = databaseHelpers.createConnection()) {
+      try (Table t1 = connection.getTable(tableName2)) {
+        int id = 0;
+        for (int i = 0; i < databaseEntriesCount / 100; i++) {
+          List<Delete> deletes = new ArrayList<>();
+          for (int j = 0; j < 100; j++) {
+            byte[] rowKey = rowKeyFromId(i);
+            deletes.add(Helpers.createDelete(rowKey, columnFamily1, qualifier1));
+            id++;
+          }
+          t1.delete(deletes);
+          assertThat(deletes).isEmpty();
+        }
+      }
+    }
+    databaseHelpers.verifyTableConsistency(tableName2);
+  }
+
+  @Test
+  public void testDeleteWithPrimaryErrors() throws IOException {
+    Assume.assumeTrue(
+        ConfigurationHelper.isPrimaryHBase() && ConfigurationHelper.isUsingHBaseMiniCluster());
+
+    int databaseEntriesCount = 1000;
+
+    // Fill tables before forcing operations to fail.
+    final TableName tableName1 = connectionRule.createTable(columnFamily1);
+    databaseHelpers.fillTable(tableName1, databaseEntriesCount, columnFamily1, qualifier1);
+
+    final TableName tableName2 = connectionRule.createTable(columnFamily1);
```

Comment:
It's not so easy, we have to fill in the tables before errors are injected before into the region implementation, otherwise some of the writes wouldn't succeed.